README.md
# flutter_sliver_tracker

滑动曝光埋点框架，支持SliverList、SliverGrid，以及嵌套的情况（详见example2/SliverMultiBoxScrollListenerDebounceDemo）

## 什么是滑动曝光埋点

滑动曝光埋点用于滑动列表组件中的模块曝光，例如Flutter中的`SliverList`、`SliverGrid`。
当`SliverList`中的某一个行（或列）移动到`ViewPort`中，并且显示比例超过一定阈值时，我们把这个事件记为一次滑动曝光事件。

当然我们对滑动曝光有一些额外的要求：
- 需要滑出一定比例的时候才出发曝光（已实现）
- 滑动速度快时不触发曝光事件（已实现）
- 滑出视野的模块，再次滑入视野时需要再次上报（已实现）
- 模块在视野中上下反复移动只触发一次曝光（已实现）
- 嵌套情况：支持滑动组件相互全套（已实现）
    - SliverMultiBoxScrollListener已支持嵌套
    - SliverMultiBoxScrollListenerDebounce已支持嵌套
    - SliverScrollListener不支持，请使用SliverMultiBoxScrollListener
    - SliverScrollListenerDebounce不支持，请使用SliverMultiBoxScrollListenerDebounce
    - 所有嵌套情况只测试过两层嵌套，理论上支持更多层。

## 运行Demo
<img src="https://raw.githubusercontent.com/SBDavid/flutter_sliver_tracker/master/demo.gif" width="270" height="480" alt="图片名称">

- 克隆代码到本地: git clone git@github.com:SBDavid/flutter_sliver_tracker.git
- 切换工作路径: cd flutter_sliver_tracker/example/
- 启动模拟器
- 运行: flutter run

## 内部原理

滑动曝光的核心难点是计算组件的露出比例。也是说我们需要知道`ListView`中的组件的`总高度`、`当前显示高度`。
这两个高度做除法就可以得出比例。

### 组件总高度
组件的总高度可以在`renderObject`中获取。我们可以获取`renderObject`下的`size`属性，其中包含了组件的长宽。

### 当前显示高度
显示高度可以从`SliverGeometry.paintExtent`中获取。

## 使用文档
### 1. 安装

```yaml
dependencies:
  flutter_sliver_tracker: ^1.0.0
```

### 2. 引用插件
```dart
import 'package:xm_sliver_listener/flutter_sliver_tracker.dart';
```

### 3. 发送滑动埋点事件

#### 3.1 通过`ScrollViewListener`捕获滚动事件，`ScrollViewListener`必须包裹在`CustomScrollView`之上。

```dart
class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      // 通过ScrollViewListener捕获滚动事件
      body: ScrollViewListener(
        child: CustomScrollView(
          slivers: <Widget>[
          ],
        ),
      ),
    );
  }
}
```

#### 3.2 在`SliverToBoxAdapter`中监听滚动停止事件，并计算显示比例
```dart
class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      // 通过ScrollViewListener捕获滚动事件
      body: ScrollViewListener(
        child: CustomScrollView(
          slivers: <Widget>[
            SliverToBoxAdapter(
              // 监听停止事件，如果在页面上展示比例，可以自行setState
              child: SliverEndScrollListener(
                onScrollInit: (SliverConstraints constraints, SliverGeometry geometry) {
                  // 显示高度 / sliver高度
                  Fluttertoast.showToast(msg: "展示比例：${geometry.paintExtent / geometry.scrollExtent}");
                },
                onScrollEnd: (ScrollEndNotification notification,
                    SliverConstraints constraints,
                    SliverGeometry geometry) {
                  Fluttertoast.showToast(msg: "展示比例：${geometry.paintExtent / geometry.scrollExtent}");
                },
                child: Container(
                  height: 300,
                  color: Colors.amber,
                 
                  ),
                ),
            ),
          ],
        ),
      ),
    );
  }
}
```

#### 3.3 在`SliverList`和`SliverGrid`中监听滚动停止事件，并计算显示比例

- itemLength：列表项布局高度
- displayedLength：列表项展示高度
- 如果需要在widget中显示高度，可以自行setState

```dart
class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      // 通过ScrollViewListener捕获滚动事件
      body: ScrollViewListener(
        child: CustomScrollView(
          slivers: <Widget>[
            SliverList(
              delegate: SliverChildBuilderDelegate(
                (BuildContext context, int index) {
                  // 监听滚动停止
                  return SliverMultiBoxScrollEndListener(
                    debounce: 1000,
                    child: Container(
                      height: 300,
                      color: Colors.redAccent,
                      child: Center(
                        child: Text("SliverList Item", style: TextStyle(fontSize: 30, color: Colors.white))
                      ),
                    ),
                    onScrollInit: (double itemLength, double displayedLength) {
                      Fluttertoast.showToast(msg: "显示高度：${displayedLength}");
                    },
                    onScrollEnd: (double itemLength, double displayedLength) {
                      Fluttertoast.showToast(msg: "显示高度：${displayedLength}");
                    },
                  );
                },
                childCount: 1
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

#### 3.4 在`SliverList`和`SliverGrid`中监听滚动更新事件，并计算显示比例
```dart
class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      // 通过ScrollViewListener捕获滚动事件
      body: ScrollViewListener(
        child: CustomScrollView(
          slivers: <Widget>[
            SliverList(
              delegate: SliverChildBuilderDelegate(
                      (BuildContext context, int index) {
                // 监听滚动更新事件
                return SliverMultiBoxScrollUpdateListener(
                  onScrollInit: (double percent) {
                    // percent 列表项显示比例
                  },
                  onScrollUpdate: (double percent) {
                    // percent 列表项显示比例
                  },
                  debounce: 1000,
                  // percent 列表项显示比例
                  builder: (BuildContext context, double percent) {
                    return Container(
                      height: 200,
                      color: Colors.amber.withAlpha((percent * 255).toInt()),
                      child: Center(
                          child: Text("SliverList Item Percent ${percent.toStringAsFixed(2)}", style: TextStyle(fontSize: 30, color: Colors.white))
                      ),
                    );
                  },
                );
              },
              childCount: 6
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

CHANGELOG.md
## [1.0.0] - 首次提交
## [1.0.1] - 更新图片地址
## [1.0.2] - bug修复：移动是context为null
## [1.0.3] - bug修复：移动是context为null
## [1.0.4] - 修改描述
## [1.0.5] - 增加滑动埋点的描述，原理说明
## [1.0.6] - 修改HomePage
## [2.0.0] - 滑动速度快时不触发曝光事件，滑出视野的模块，再次滑入视野时需要再次上报，模块在视野中上下反复移动只触发一次曝光
## [2.0.1] - 在首帧后出发曝光事件 addPostFrameCallback
## [2.0.2] - 优化：删除多余的print
## [2.0.5] - 优化：在dispose中添加异常处理
## [2.0.6] - 增加接口：暴露滚动事件
## [2.0.7] - 优化：支持滑动列表的嵌套
## [3.0.0] - 新增功能：支持嵌套形式的滑动曝光埋点
## [3.0.1] - bugfix: 查询父节点曝光情况
## [3.0.2] - bugfix: debounce中判断显示比例
## [3.0.3] - bugfix: 当元素处于PageView中时，可能执行initState但是实际不发生渲染，并且context为null
## [3.0.4] - bugfix: displayedLength默认值为0
## [3.0.5] - bugfix: 列表初始化后没有暴露在可视范围内

example/lib/mainB.dart
import 'package:flutter/material.dart';
import 'package:flutter_sliver_tracker/flutter_sliver_tracker.dart';
import 'package:fluttertoast/fluttertoast.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'SliverListenerDemo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'SliverListenerDemo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(

        title: Text(widget.title),
      ),
      body: ScrollViewListener(
        child: CustomScrollView(
          scrollDirection: Axis.vertical,
          slivers: <Widget>[
            SliverAppBar(
              title: Text("Listen Scroll End in SliverToBoxAdapter"),
            ),

            SliverToBoxAdapter(
              child: SliverScrollListener(
                onScrollInit: (SliverConstraints constraints, SliverGeometry geometry) {
                  // Fluttertoast.showToast(msg: "展示比例：${geometry.paintExtent / geometry.scrollExtent}");
                },
                onScrollEnd: (ScrollEndNotification notification,
                    SliverConstraints constraints,
                    SliverGeometry geometry) {
                  // Fluttertoast.showToast(msg: "展示比例：${geometry.paintExtent / geometry.scrollExtent}");
                },
                child: Container(
                  height: 300,
                  color: Colors.amber,
                  child: Center(
                    child: Text("Pull to show paintExtent", style: TextStyle(fontSize: 30, color: Colors.white),),
                  ),
                ),
              ),
            ),

            SliverAppBar(
              title: Text("Listen Scroll End in SliverList"),
            ),
            SliverList(
              delegate: SliverChildBuilderDelegate(
                (BuildContext context, int index) {
                  return SliverMultiBoxScrollEndListener(
                    debounce: 1000,
                    child: Container(
                      height: 300,
                      color: Colors.redAccent,
                      child: Center(
                        child: Text("SliverList Item", style: TextStyle(fontSize: 30, color: Colors.white))
                      ),
                    ),
                    onScrollInit: (double itemLength, double displayedLength) {
                      //Fluttertoast.showToast(msg: "显示高度：${displayedLength}");
                    },
                    onScrollEnd: (double itemLength, double displayedLength) {
                      //Fluttertoast.showToast(msg: "显示高度：${displayedLength}");
                    },
                  );
                },
                childCount: 1
              ),
            ),
            SliverAppBar(
              title: Text("Listen Scroll End in SliverGrid"),
            ),
            SliverGrid(
              gridDelegate: SliverGridDelegateWithMaxCrossAxisExtent(
                maxCrossAxisExtent: 200.0,
                mainAxisSpacing: 10.0,
                crossAxisSpacing: 10.0,
                childAspectRatio: 1.0,
              ),
              delegate: SliverChildBuilderDelegate(
                    (BuildContext context, int index) {
                  return SliverMultiBoxScrollEndListener(
                    child: Container(
                      alignment: Alignment.center,
                      color: Colors.teal[100 * (index % 9)],
                      child: Text('Grid Item $index'),
                    ),
                    onScrollInit: (double itemLength, double displayedLength) {
                      Fluttertoast.showToast(msg: "显示高度：$displayedLength");
                    },
                    onScrollEnd: (double itemLength, double displayedLength) {
                      Fluttertoast.showToast(msg: "显示高度：$displayedLength");
                    },
                  );
                },
                childCount: 2,
              ),
            ),
            SliverAppBar(
              title: Text("Listen Scroll Update in SliverList"),
              pinned: true,
            ),
            SliverList(
              delegate: SliverChildBuilderDelegate(
                      (BuildContext context, int index) {
                    return SliverMultiBoxScrollUpdateListener(
                      onScrollInit: (double percent) {
                        // percent 列表项显示比例
                      },
                      onScrollUpdate: (double percent) {
                        // percent 列表项显示比例
                      },
                      debounce: 1000,
                      builder: (BuildContext context, double percent) {
                        return Container(
                          height: 200,
                          color: Colors.amber.withAlpha((percent * 255).toInt()),
                          child: Center(
                              child: Text("SliverList Item Percent ${percent.toStringAsFixed(2)}", style: TextStyle(fontSize: 30, color: Colors.white))
                          ),
                        );
                      },
                    );
                  },
                  childCount: 6
              ),
            ),
            SliverAppBar(
              title: Text("End"),
            ),
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
